src/core/Akka/Serialization/Serialization.cs

```diff
@@ -229,20 +229,34 @@ public Serializer FindSerializerFor(object obj)
         /// <returns>TBD</returns>
         public Serializer FindSerializerForType(Type objectType)
         {
-            Type type = objectType;
-            //TODO: see if we can do a better job with proper type sorting here - most specific to least specific (object serializer goes last)
-            foreach (var serializerType in _serializerMap)
+            Serializer fullMatchSerializer;
+            if (_serializerMap.TryGetValue(objectType, out fullMatchSerializer))
             {
-                //force deferral of the base "object" serializer until all other higher-level types have been evaluated
-                if (serializerType.Key.IsAssignableFrom(type) && serializerType.Key != _objectType)
-                    return serializerType.Value;
+                return fullMatchSerializer;
             }
+            else
+            {
+                Serializer serializer = null;
+                Type type = objectType;
+
+                // TODO: see if we can do a better job with proper type sorting here - most specific to least specific (object serializer goes last)
+                foreach (var serializerType in _serializerMap)
+                {
+                    // force deferral of the base "object" serializer until all other higher-level types have been evaluated
+                    if (serializerType.Key.IsAssignableFrom(type) && serializerType.Key != _objectType)
+                        serializer = serializerType.Value;
+                }
 
-            //do a final check for the "object" serializer
-            if (_serializerMap.ContainsKey(_objectType) && _objectType.IsAssignableFrom(type))
-                return _serializerMap[_objectType];
+                //do a final check for the "object" serializer
+                if (_serializerMap.ContainsKey(_objectType) && _objectType.IsAssignableFrom(type))
+                    serializer = _serializerMap[_objectType];
 
-            throw new SerializationException($"Serializer not found for type {objectType.Name}");
+                if (serializer == null)
+                    throw new SerializationException($"Serializer not found for type {objectType.Name}");
+
+                _serializerMap.Add(type, serializer);
```

Comment:
Why not just change it to `_serializerMap.Add[type] = serializer;` ? Becomes an upsert that way.